homeassistant/components/modbus/validators.py

```diff
@@ -73,19 +76,35 @@
 # As expressed in DEFAULT_STRUCT_FORMAT
 
 DEFAULT_STRUCT_FORMAT = {
-    DataType.INT8: ENTRY("b", 1, PARM_IS_LEGAL(False, False, False, False, False)),
-    DataType.UINT8: ENTRY("c", 1, PARM_IS_LEGAL(False, False, False, False, False)),
-    DataType.INT16: ENTRY("h", 1, PARM_IS_LEGAL(False, False, True, True, False)),
-    DataType.UINT16: ENTRY("H", 1, PARM_IS_LEGAL(False, False, True, True, False)),
-    DataType.FLOAT16: ENTRY("e", 1, PARM_IS_LEGAL(False, False, True, True, False)),
-    DataType.INT32: ENTRY("i", 2, PARM_IS_LEGAL(False, False, True, True, True)),
-    DataType.UINT32: ENTRY("I", 2, PARM_IS_LEGAL(False, False, True, True, True)),
-    DataType.FLOAT32: ENTRY("f", 2, PARM_IS_LEGAL(False, False, True, True, True)),
-    DataType.INT64: ENTRY("q", 4, PARM_IS_LEGAL(False, False, True, True, True)),
-    DataType.UINT64: ENTRY("Q", 4, PARM_IS_LEGAL(False, False, True, True, True)),
-    DataType.FLOAT64: ENTRY("d", 4, PARM_IS_LEGAL(False, False, True, True, True)),
-    DataType.STRING: ENTRY("s", -1, PARM_IS_LEGAL(True, False, False, False, False)),
-    DataType.CUSTOM: ENTRY("?", 0, PARM_IS_LEGAL(True, True, False, False, False)),
+    DataType.INT8: ENTRY(
+        "b", 1, PARM_IS_LEGAL(False, False, False, False, False, True)
+    ),
+    DataType.UINT8: ENTRY(
+        "c", 1, PARM_IS_LEGAL(False, False, False, False, False, True)
+    ),
+    DataType.INT16: ENTRY("h", 2, PARM_IS_LEGAL(False, False, True, True, False, True)),
+    DataType.UINT16: ENTRY(
+        "H", 2, PARM_IS_LEGAL(False, False, True, True, False, True)
+    ),
+    DataType.FLOAT16: ENTRY(
+        "e", 2, PARM_IS_LEGAL(False, False, True, True, False, True)
+    ),
+    DataType.INT32: ENTRY("i", 4, PARM_IS_LEGAL(False, False, True, True, True, True)),
```

Comment:
Just had a conversation with another maintainer, who pointed out something I missed.

You probably need to multiply the this table somehow, because 
- for register_size 4: int16 is not allowed
- for register_size 8: int32 is not allowed
etc.
This of course need to go into the corresponding table in the documentation.

Reply:
Why have the validation in the table when it can be done in code with simple math. Similar to how you currently validate structure size and count in custom types. Thats what i tried to do

Reply:
It is not a demand to have it in the table, history have just shown, that for maintenance reasons it is simpler to have these tests in one central place. I used to have all these checks in the validator, but the code became very unstable on not very resistant to changes, hence the change to a table. Which btw also makes it easier to document the limitations.

I have not see a check in your code that e.g. register_size=4 means int16 is not allowed...this needs to be a clear error message to the user not just that count and size do not match.

The validate structure size and count is mostly done at runtime, here we talk about catching errors when configuring the integration, but of course it can be done in the validator.